### PR TITLE
[WIP] Quick fix to provide complex data type support for hashmap based algorithms

### LINKED ISSUE #17927: Functions that rely on hash tables are incorrect for complex numbers
It looks like various functions that rely on hash tables fail when complex numbers are present, as it appears they only use the real component.

Consider the following list of values:
```python
In [2]: x = [0, 1j, 1, 1+1j, 1+2j]

In [3]: x
Out[3]: [0, 1j, 1, (1+1j), (1+2j)]
```

Using `value_counts`:
```python
In [4]: pd.value_counts(x)
Out[4]:
(1+0j)    3
0j        2
dtype: int64
```

Using `unique`:
```python
In [5]: pd.unique(x)
Out[5]: array([ 0.,  1.])
```

Using `duplicated`:
```python
In [6]: pd.Series(x).duplicated()
Out[6]:
0    False
1     True
2    False
3     True
4     True
dtype: bool
```

Using `isin`:
```python
In [7]: pd.Series(x).isin([1j, 1+1j, 1+2j])
Out[7]:
0    False
1    False
2    False
3    False
4    False
dtype: bool

In [8]: pd.Series(x).isin([0, 1])
Out[8]:
0    True
1    True
2    True
3    True
4    True
dtype: bool
```

Using `factorize` fails as described in #16399, and multiple other functions fail in similar ways (`rank`, `nunique`,  `mode`, etc.).

Note that these appear to work if the `dtype` is explicitly set to `object` instead of `complex64`:
```python
In [9]: pd.Series(x, dtype=object).isin([0, 1])
Out[9]:
0     True
1    False
2     True
3    False
4    False
dtype: bool
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 51c5f4d2a20179041114fac0262ef567d0b0373f
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.21.0rc1+26.g51c5f4d
pytest: 3.1.2
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.13.1
scipy: 0.19.1
pyarrow: 0.6.0
xarray: 0.9.6
IPython: 6.1.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.2
feather: 0.4.0
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 0.9.8
lxml: 3.8.0
bs4: None
html5lib: 0.999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: 0.1.0
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ghost**
- [x] fixes #17927
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [-] whatsnew entry

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -104,10 +104,11 @@ def _ensure_data(values, dtype=None):
...
             # which discards complex parts
-            with catch_warnings():
-                simplefilter("ignore", np.ComplexWarning)
-                values = ensure_float64(values)
-            return values, "float64", "float64"
+            # with catch_warnings():
+            #     simplefilter("ignore", np.ComplexWarning)
+            #     values = ensure_float64(values)
+            # return values, "float64", "float64"
+            return ensure_object(np.asarray(values)), "object", "object"
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py]: we cant just return non-object values?  regardless, definitely dont leave the commented-out stuff above, just delete

File: pandas/core/sorting.py [hunk 2]
```diff
@@ -444,6 +444,11 @@ def safe_sort(values, labels=None, na_sentinel=-1, assume_unique=False, verify=T
         dtype, _ = infer_dtype_from_array(values)
         values = np.asarray(values, dtype=dtype)
 
+    elif isinstance(values, np.ndarray) and values.dtype == object:
```
Comment (WillAyd) [inline pandas/core/sorting.py]: Should use `is_object_dtype` from pandas.core.dtypes.common

File: pandas/tests/test_complex.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @pytest.mark.parametrize("array,expected", [
+        (
+            [1 + 1j, 0, 1, 1j, 1 + 2j, 1 + 2j],
+            np.array([(1 + 1j), 0j, (1 + 0j), 1j, (1 + 2j)], dtype=object)
+        ),
+    ])
+    def test_unique(self, array, expected):
+        result = algos.unique(array)
+        assert np.array_equal(result, expected)
```
Comment (WillAyd) [inline pandas/tests/test_complex.py]: Can use `tm.assert_numpy_array_equal`

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/test_complex.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            [1 + 2j, 0, 1j, 1, 1j, 1 + 1j],
+            # index is sorted by value counts in descending order by default
+            Series([2, 1, 1, 1, 1], index=[1j, 1 + 2j, 1 + 1j, 1, 0])
+        )
+    ])
+    def test_value_counts(self, array, expected):
+        result = algos.value_counts(array)
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize("array,expected", [
```
Comment (WillAyd) [inline pandas/tests/test_complex.py]: For any that are just one set of parameters here you can just define in the function body. No need to parametrize

File: pandas/tests/test_complex.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        (
+            [1, 2, 2 + 1j],
+            (np.array([0, 1, 2], dtype=np.int64),
+             np.array([(1 + 0j), (2 + 0j), (2 + 1j)], dtype=object))
+        ),
+    ])
+    def test_factorize(self, array, expected):
+        result = pd.factorize(array)
+        assert len(result) == 2
```
Comment (WillAyd) [inline pandas/tests/test_complex.py]: I think easier to read at a quick glance if you just compare the two elements rather than the enumerating loop

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/test_complex.py (hunk 4 above)
Comment (heckeop) [reply pandas/tests/test_complex.py]: This is just an initial set of test cases. I plan on adding more test cases for each of the algos.

**Comment by heckeop** [comment]
Still active. I am on a vacation at the moment, but will be back on this
shortly.

**Comment by jreback** [comment]
@heckeop can you merge master and update

**Comment by heckeop** [comment]
has the underlying numpy undergone a version change? The test below was passing before. Now I am getting type conflicts.

https://github.com/pandas-dev/pandas/pull/27599/files#diff-3957a56d910fcc0790307c20a9a67a58R29-R40

File: pandas/core/algorithms.py [hunk 6]
```diff
@@ -101,13 +101,7 @@ def _ensure_data(values, dtype=None):
...
             return ensure_object(np.asarray(values)), "object", "object"
         elif is_complex_dtype(values) or is_complex_dtype(dtype):
-
-            # ignore the fact that we are casting to float
-            # which discards complex parts
-            with catch_warnings():
-                simplefilter("ignore", np.ComplexWarning)
-                values = ensure_float64(values)
-            return values, "float64", "float64"
+            raise TypeError("Complex data types not supported...Coercing to object")
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py]: better to do the object-coersion here than fallback to the except branch.  clearer

**Comment by jbrockmendel** [comment]
> has numpy undergone a version change? The test below was passing before. Now, the returned object has dtype complex128

We've made some recent changes to the cython code that handled complex correctly in some cases, not sure if this would be affected.  can you merge master and re-push

**Comment by alimcmaster1** [comment]
Thanks for the PR - @heckeop Could you merge master here?

**Comment by jbrockmendel** [comment]
@heckeop can you rebase and see if you can get the CI passing

**Comment by jreback** [comment]
@heckeop can you merge master

**Comment by alimcmaster1** [comment]
I can maybe take a look if we don’t hear back from @heckeop

**Comment by WillAyd** [comment]
Closing as I think this is stale but @heckeop ping if you'd like to pick back up

**Pull request closed without merging**